sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs

```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections;
+
+namespace System.ClientModel.Primitives;
+
+internal abstract class CollectionWriter
+{
+    internal static CollectionWriter GetCollectionWriter(IEnumerable enumerable, ModelReaderWriterOptions options)
+    {
+        //For info on the different formats see the comments in ModelReaderWriterOptions.cs
+        if (options.Format != "J" && options.Format != "W")
+        {
+            throw new InvalidOperationException($"Format '{options.Format}' is not supported.  Only 'J' or 'W' format can be written as collections");
+        }
+
+        if (options.Format == "J")
+        {
+            return new JsonCollectionWriter();
+        }
+        else // W format
+        {
+            var persistableModel = GetIPersistableModel(enumerable);
+            var wireFormat = persistableModel.GetFormatFromOptions(options);
+            if (wireFormat == "J" && persistableModel is IJsonModel<object>)
+            {
+                return new JsonCollectionWriter();
+            }
+            throw new InvalidOperationException($"{persistableModel.GetType().Name} has a wire format of '{wireFormat}'.  It must be 'J' to be written as a collection");
+        }
+    }
+
+    private static IPersistableModel<object> GetIPersistableModel(IEnumerable enumerable)
+    {
+        object first = GetFirstObject(enumerable);
+        if (first is IEnumerable nextEnumerable)
+        {
+            return GetIPersistableModel(nextEnumerable);
+        }
+        else if (first is IPersistableModel<object> persistableModel)
+        {
+            return persistableModel;
+        }
+        else
+        {
+            throw new InvalidOperationException($"Unable to write {enumerable.GetType().Name}.  Only collections of 'IPersistableModel' can be written.");
+        }
+    }
+
+    private static object GetFirstObject(IEnumerable enumerable)
+    {
+        if (enumerable is IDictionary dictionary)
```

Comment:
Given a `Dictionary<foo, object>`, where some of the objects are `IEnumerable` and some are not, what would happen?

Reply:
Added a test to show how you can make a mix work.  It requires you to write your own IJsonModel discriminator.  Writing mixed works with no custom discriminator but reading only works with.